doc/developer/design/20230802_webhook_idempotency.md

```diff
@@ -0,0 +1,403 @@
+provide a sender and receiver for requests that can both register and
+un-register `GlobalId`s to bidirectional channels.
+
+In essence, this gRPC server works as a reverse proxy for sending requests to
+Timely workers.
+
+**Early-arriving messages**
+
+Each `clusterd` will spawn a webhook server on boot, just as it does with the
+storage server.
+
+This means that there is possibility of webhook requests coming into the service
+because it's been given a channel to propagate them to. In this case, I suggest
+we just return a status code equivalent to service not available yet, please
+retry later.
+
+**`mz_service`?**
+
+`mz_service`'s module-level documentation says:
+
+> The code shared here is not obviously reusable by future services we may split
+> out of environmentd.
+
+And poking around the code, this seems to be true. Paul and I noodled on how to
+use this module for this service and while it seems possible, it seems more
+legible and simpler to maintain to just roll a new gRPC server.
+
+During the implementation, I plan to use persist's PubSub service as a reference
+implementation and will extract common code from it into some _other_ shard
+module and/or unify the implementation with `mz_service`.
+
+#### Storage server + `clusterd`
+
+`clusterd` will house the new webhook gRPC server.
+
+We will spin up the webhook gRPC server before the storage server.
+
+The storage server will grow a handle to the webhook server, which will be able
+to handle registering and un-registering webhook sources. Commands to register a
+source will provide a `GlobalId` and, in response, get handle to a stream to
+read requests from, as well as send responses to. This handle will eventually
+end up as part of a worker's `StorageState`, and the channels will end up in the
+dataflow for the webhook source.
+
+#### `environmentd` Handler
+
+Some endpoint in `environmentd` will grow the ability to send gRPC requests to
+the webhook service running on each cluster. This bit of code will need the
+ability to understand which cluster each source is running on, as well as have a
+means of understanding (and likely caching) which port on the cluster to find
+the webhook service.
+
+For simplicity's sake, we will still have this endpoint perform validation and
+extract the request's idempotency token.
+
+However, we will now have it also stash the channel to respond to in a
+`BTreeMap<IdempotencyToken, ResponseChannel>`. As it receives responses from the
+webhook gRPC service, it will send the appropriate response along to the
+awaiting client. Responses will be akin to some set of:
+
+- `200` for successes (response from webhook server)
+- `404` if the webhook server doesn't yet have a channel set up for the source
+  yet (response from webhook server)
+- `408` if the webhook server doesn't respond (response from handler)
+- `502` if `environmentd` doesn't believe the request is for a webhook source
+  (response from handler)
```

Comment:
This one seems like an internal server error? Or am I misunderstanding the condition that's being checked?

```suggestion
- `500` if `environmentd` doesn't believe the request is for a webhook source
  (response from handler)
```